src/Interpreters/Cache/FileCache.h

```diff
@@ -71,252 +76,177 @@ using QueryContextPtr = std::shared_ptr<QueryContext>;
+        RETURN_NULL,
+    };
 
-        size_t size() const { return file_segment->reserved_size; }
+    LockedKeyPtr createLockedKey(const Key & key, KeyNotFoundPolicy key_not_found_policy);
 
-        FileSegmentCell(FileSegmentPtr file_segment_, FileCache * cache, std::lock_guard<std::mutex> & cache_lock);
+    mutable KeysQueuePtr cleanup_keys_metadata_queue;
 
-        FileSegmentCell(FileSegmentCell && other) noexcept
-            : file_segment(std::move(other.file_segment)), queue_iterator(std::move(other.queue_iterator)) {}
-    };
+    FileCachePriorityPtr main_priority;
+    mutable CacheGuard cache_guard;
 
-    using AccessKeyAndOffset = std::pair<Key, size_t>;
-    struct KeyAndOffsetHash
+    struct HitsCountStash
     {
-        std::size_t operator()(const AccessKeyAndOffset & key) const
+        HitsCountStash(size_t hits_threashold_, size_t queue_size_)
+            : hits_threshold(hits_threashold_), queue(std::make_unique<LRUFileCachePriority>(0, queue_size_))
         {
-            return std::hash<UInt128>()(key.first.key) ^ std::hash<UInt64>()(key.second);
+            if (!queue_size_)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Queue size for hits queue must be non-zero");
         }
-    };
-
-    using FileSegmentsByOffset = std::map<size_t, FileSegmentCell>;
-    using CachedFiles = std::unordered_map<Key, FileSegmentsByOffset>;
-    using FileCacheRecords = std::unordered_map<AccessKeyAndOffset, IFileCachePriority::WriteIterator, KeyAndOffsetHash>;
 
-    CachedFiles files;
-    std::unique_ptr<IFileCachePriority> main_priority;
-
-    FileCacheRecords stash_records;
-    std::unique_ptr<IFileCachePriority> stash_priority;
-    size_t max_stash_element_size;
-
-    void loadCacheInfoIntoMemory(std::lock_guard<std::mutex> & cache_lock);
+        const size_t hits_threshold;
+        FileCachePriorityPtr queue;
+        using Records = std::unordered_map<KeyAndOffset, IFileCachePriority::Iterator, FileCacheKeyAndOffsetHash>;
+        Records records;
+    };
 
-    FileSegments getImpl(const Key & key, const FileSegment::Range & range, std::lock_guard<std::mutex> & cache_lock);
+    mutable std::unique_ptr<HitsCountStash> stash;
 
-    FileSegmentCell * getCell(const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock);
+    FileCacheQueryLimitPtr query_limit;
 
-    /// Returns non-owned pointer to the cell stored in the `files` map.
-    /// Doesn't reserve any space.
-    FileSegmentCell * addCell(
-        const Key & key,
-        size_t offset,
-        size_t size,
-        FileSegment::State state,
-        const CreateFileSegmentSettings & create_settings,
-        std::lock_guard<std::mutex> & cache_lock);
+    void assertInitialized() const;
 
-    static void useCell(const FileSegmentCell & cell, FileSegments & result, std::lock_guard<std::mutex> & cache_lock);
+    void loadMetadata();
 
-    bool tryReserveForMainList(
+    FileSegments getImpl(
         const Key & key,
```

Comment:
We shouldn't pass `Key` parameter if there is `LockedKey` parameter. Technically, it allows modifying key metadata w/o lock. So, having `Key` parameter excludes `LockedKey` parameter in method declaration and vice versa

Reply:
> Technically, it allows modifying key metadata w/o lock.

arguable conclusion, technically this can be said about any key in cache, so it does not mean that something leaks if I pass key as a separate argument.

But I agree that we can omit passing key if there is LockedKey argument. Though I personally like that key is passed as a separate argument because it is a clear pattern of these cache methods to have the same arguements: `get/set/getOrSet/getImpl(key, offset, ... additional args ...)`;

Reply:
Well, what happens if `Key` and `LockedKey` point to different keys? When in a method's signature, `Key` and `LockedKey` exclude each other - there is no such question.

Reply:
After another thought I think it is a great idea, btw there are actually not just one but several methods where this change needs to be done